File: src/util/paths.ts
import * as os from "node:os";
import * as path from "node:path";
import * as fs from "node:fs";

export const IS_WIN = process.platform === "win32";
export const IS_MAC = process.platform === "darwin";

function resolveHome(): string {
  return process.env.HOME || os.homedir();
}

export function appDataDir(): string {
  const h = resolveHome();
  if (IS_WIN) return process.env.APPDATA || path.join(h, "AppData", "Roaming");
  if (IS_MAC) return path.join(h, "Library", "Application Support");
  return process.env.XDG_CONFIG_HOME || path.join(h, ".config");
}

let HOME_OVERRIDE: string | null = null;
export function setHomeOverride(p: string | null) {
  HOME_OVERRIDE = p;
}
export function home(): string {
  return HOME_OVERRIDE ?? resolveHome();
}
export function configRoot(): string {
  if (HOME_OVERRIDE) {
    if (IS_WIN) return path.join(HOME_OVERRIDE, "AppData", "Roaming");
    if (IS_MAC) return path.join(HOME_OVERRIDE, "Library", "Application Support");
    return path.join(HOME_OVERRIDE, ".config");
  }
  return appDataDir();
}

export function ensureDir(p: string): void {
  fs.mkdirSync(p, { recursive: true });
}

export function readFileSafe(p: string): string | null {
  try {
    return fs.readFileSync(p, "utf8");
  } catch {
    return null;
  }
}

export function writeFile(p: string, content: string): void {
  ensureDir(path.dirname(p));
  fs.writeFileSync(p, content, "utf8");
}

export function exists(p: string): boolean {
  try {
    fs.accessSync(p);
    return true;
  } catch {
    return false;
  }
}

// Agent-specific config locations resolved against current home/configRoot.
export const agentPaths = {
  claudeCode: () => ({
    dir: path.join(home(), ".claude"),
    settings: path.join(home(), ".claude", "settings.json"),
    globalJson: path.join(home(), ".claude.json"), // mcpServers live here for user scope
    instructions: path.join(home(), ".claude", "CLAUDE.md"),
    skillsDir: path.join(home(), ".claude", "skills"),
  }),
  opencode: () => {
    const dir = path.join(configRoot(), "opencode");
    const candidates = [
      path.join(dir, "opencode.jsonc"),
      path.join(dir, "opencode.json"),
      path.join(dir, "config.json"),
    ];
    const existing = candidates.find((c) => {
      try { fs.accessSync(c); return true; } catch { return false; }
    });
    return {
      dir,
      config: existing ?? path.join(dir, "opencode.jsonc"),
      instructions: path.join(dir, "AGENTS.md"),
      pluginsDir: path.join(dir, "plugins"),
      rulesDir: path.join(dir, "rules"),
    };
  },
  codex: () => {
    const envHome = process.env.CODEX_HOME;
    const dir = envHome && envHome.length > 0 ? envHome : path.join(home(), ".codex");
    return {
      dir,
      config: path.join(dir, "config.toml"),
      instructions: path.join(dir, "AGENTS.md"),
    };
  },
};

File: internal/util/paths.go
package util

import (
	"os"
	"path/filepath"
	"runtime"
)

var (
	IsWin = runtime.GOOS == "windows"
	IsMac = runtime.GOOS == "darwin"
)

var homeOverride string

// SetHomeOverride redirects all home-relative paths (used by tests/sandbox).
func SetHomeOverride(p string) { homeOverride = p }

func resolveHome() string {
	if h := os.Getenv("HOME"); h != "" {
		return h
	}
	h, _ := os.UserHomeDir()
	return h
}

func Home() string {
	if homeOverride != "" {
		return homeOverride
	}
	return resolveHome()
}

func AppDataDir() string {
	h := resolveHome()
	if IsWin {
		if a := os.Getenv("APPDATA"); a != "" {
			return a
		}
		return filepath.Join(h, "AppData", "Roaming")
	}
	if IsMac {
		return filepath.Join(h, "Library", "Application Support")
	}
	if x := os.Getenv("XDG_CONFIG_HOME"); x != "" {
		return x
	}
	return filepath.Join(h, ".config")
}

func ConfigRoot() string {
	if homeOverride != "" {
		if IsWin {
			return filepath.Join(homeOverride, "AppData", "Roaming")
		}
		if IsMac {
			return filepath.Join(homeOverride, "Library", "Application Support")
		}
		return filepath.Join(homeOverride, ".config")
	}
	return AppDataDir()
}

func EnsureDir(p string) error { return os.MkdirAll(p, 0o755) }

func ReadFileSafe(p string) (string, bool) {
	b, err := os.ReadFile(p)
	if err != nil {
		return "", false
	}
	return string(b), true
}

func WriteFile(p, content string) error {
	if err := EnsureDir(filepath.Dir(p)); err != nil {
		return err
	}
	return os.WriteFile(p, []byte(content), 0o644)
}

func Exists(p string) bool {
	_, err := os.Stat(p)
	return err == nil
}

// ClaudePaths holds Claude Code config locations.
type ClaudePaths struct {
	Dir, Settings, GlobalJSON, Instructions, SkillsDir string
}

func ClaudeCodePaths() ClaudePaths {
	h := Home()
	return ClaudePaths{
		Dir:          filepath.Join(h, ".claude"),
		Settings:     filepath.Join(h, ".claude", "settings.json"),
		GlobalJSON:   filepath.Join(h, ".claude.json"),
		Instructions: filepath.Join(h, ".claude", "CLAUDE.md"),
		SkillsDir:    filepath.Join(h, ".claude", "skills"),
	}
}

// OpenCodePaths holds OpenCode config locations.
type OpenCodePaths struct {
	Dir, Config, Instructions, PluginsDir, RulesDir string
}

func OpenCodePathsResolved() OpenCodePaths {
	dir := filepath.Join(ConfigRoot(), "opencode")
	candidates := []string{
		filepath.Join(dir, "opencode.jsonc"),
		filepath.Join(dir, "opencode.json"),
		filepath.Join(dir, "config.json"),
	}
	config := filepath.Join(dir, "opencode.jsonc")
	for _, c := range candidates {
		if Exists(c) {
			config = c
			break
		}
	}
	return OpenCodePaths{
		Dir:          dir,
		Config:       config,
		Instructions: filepath.Join(dir, "AGENTS.md"),
		PluginsDir:   filepath.Join(dir, "plugins"),
		RulesDir:     filepath.Join(dir, "rules"),
	}
}

// CodexPaths holds Codex config locations.
type CodexPaths struct {
	Dir, Config, Instructions string
}

func CodexPathsResolved() CodexPaths {
	dir := filepath.Join(Home(), ".codex")
	if env := os.Getenv("CODEX_HOME"); env != "" {
		dir = env
	}
	return CodexPaths{
		Dir:          dir,
		Config:       filepath.Join(dir, "config.toml"),
		Instructions: filepath.Join(dir, "AGENTS.md"),
	}
}
